### PR TITLE
Grids now have a topology

### LINKED ISSUE #446: Use `FT` for "floating point type"
In various parts of the code we use either `T` or `FT` to indicate the floating point type (eg, `Float64` or `Float32`).

We should use `FT` consistently, as `CLIMA` will:

https://github.com/climate-machine/CLIMA/pull/461

### LINKED ISSUE #459: Use less defaults in model construction?
@johncmarshall54 suggested it might be a bit dangerous to assume default boundary conditions, so they should always be specified.

Worth discussing as this might apply to other model properties too.

A common question for simple model setups is "what are the boundary conditions?".

### LINKED ISSUE #489: Grid `Topology` concept
I think the concept of grid `Topology` will be useful.

The `Topology` codifies information about the domain in which the grid is defined. We have three (currently implicit) topologies for each orthogonal direction `x`, `y`, `z`: `Flat`, `Bounded`, and `Periodic`. 

A `Flat` topology has `N=1`. A `Bounded` topology has boundaries (either closed or open, though we only supported closed boundaries right now) on left and right. A `Periodic` topology is periodic in the specified direction.

I propose that we create 3 new type parameters of the grid that codify whether the grid is `Flat`, `Bounded`, or `Periodic` in each of `x`, `y`, `z`. 

We can add a keyword to the constructors for grids that allow users to specify the grid topology. For example, the topology that we call a 'Channel Model' would be specified via

```julia
grid = RegularCartesianGrid(size=(64, 64, 64), length=(1, 1, 1), topology=(Periodic, Bounded, Bounded)).
```

A two-dimensional grid can also be defined.

Topologies imply default boundary conditions: a `Bounded` direction has no-flux boundary conditions; a `Periodic` direction has `Periodic` boundary conditions. For a `Flat` direction we can also use no flux conditions, though it should be irrelevant. 

This change will eliminate the need for a special `ChannelModel` constructor, and also eliminates the need for constructors with names like `HorizontallyPeriodicBCs`. Instead, we have a generic constructor for `FieldBoundaryConditions` that lets a user specify boundary conditions in any direction. User-supplied boundary conditions are then checked for consistency with the specified topology, and missing default boundary conditions are added where none exist.

A `Flat` topology will enable us to set halo sizes to 0 for `Flat` directions, and elide the unnecessary interpolation and differentiation operations in `Flat` directions that currently plagues two and one-dimensional models.

A related issue is #330, but this proposal has wider-ranging consequences so I decided to open a new issue.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR adds grid topologies: `Periodic`, `Bounded`, and `Singleton`. We should finalize our choice of named before merging.

As `topology` is now a required kwarg of all grid constructors tons of refactoring was needed. It also does a bit of cleanup: `AbstractGrid` is now defined in the Grids submodule, and grids no longer have the `Tx, Ty, Tz` property (total number of grid points).

Documentation should be updated before this PR is merged.

Resolves #446 
Resolves #459
Resolves #489

### CODE REVIEW

**Comment by glwagner** [comment]
I don’t like “Singleton”: Periodic and Bounded refer to the physical nature of the dimension, while Singleton obliquely references the size of an underlying array of discrete data. A better name would refer to the physical nature of the coordinate; “Uniform” or “Flat” are examples. 

It could make sense to make (Periodic, Periodic, Bounded) a default. This is already the default of Model.

**Comment by ali-ramadhan** [comment]
> I don’t like “Singleton”: Periodic and Bounded refer to the physical nature of the dimension, while Singleton obliquely references the size of an underlying array of discrete data. A better name would refer to the physical nature of the coordinate; “Uniform” or “Flat” are examples.

We're setting up a discrete model anyways, but I guess if it's supposed to model some physical system then "Singleton" may not be completely appropriate.

I'd vote for Flat over Uniform then. Uniform could be multiple grid points while Flat strongly implies one grid point (while both imply homogenous dynamics in that dimension).

> It could make sense to make (Periodic, Periodic, Bounded) a default. This is already the default of Model.

Yeah I guess we have a choice here: do we want to make horizontally periodic the default or are users now required to specify a grid topology?

I could see benefits either way. Keeping it the default would introduce no breaking changes and simplify the tests, but requiring it would improve the clarity of most scripts (an issue that has been brought up before, see #459). I think most real scripts (not tests) should be specifying a topology for clarity.

**Comment by ali-ramadhan** [comment]
Another design we can consider is something like:
```julia
grid = RegularCartesianGrid(size=(x=128, z=64), x=(-1, 1), z=(-1, 0), topology=(x=Periodic, z=Bounded))
```
where since `y` has not been specified, it is assumed to be `Flat`.

This makes the grid constructors more verbose but has the added benefit of being clearer and not having to specify flat dimensions, i.e. it's an xz-model so y should not even be mentioned.

It would require a lot more refactoring though... but now's the time I guess.

**Comment by glwagner** [comment]
> It would require a lot more refactoring though... but now's the time I guess.

I think explicit is better.

Why do we need to refactor? Why don't we just specify a default topology in the grid constructor? Then, the only refactoring we need to do is for models in the channel domain.

**Comment by glwagner** [comment]
> I'd vote for Flat over Uniform then. Uniform could be multiple grid points while Flat strongly implies one grid point (while both imply homogenous dynamics in that dimension).

I'm fine with `Flat`. But the topology does not restrict how many grid points one uses. I don't see the point of doing that. We only need this for eliding derivative operations, not restricting user behavior.

**Comment by ali-ramadhan** [comment]
> Why do we need to refactor? Why don't we just specify a default topology in the grid constructor? Then, the only refactoring we need to do is for models in the channel domain.

I'll add a default `topology=(Periodic, Periodic, Bounded)` then.

The bigger refactor would be if we decided to go with named tuples like `size=(x=16, z=32)` instead of just tuples `size=(16, 1, 32)`.

> I'm fine with `Flat`.

`Flat` it is then.

> But the topology does not restrict how many grid points one uses. I don't see the point of doing that. We only need this for eliding derivative operations, not restricting user behavior.

True but I don't see the point of running with more than one grid point in a dimension that's supposed to be `Singleton` or `Flat`. But I guess users can do many weird things, and this wouldn't be the weirdest.

File: src/Buoyancy/Buoyancy.jl [hunk 1]
```diff
@@ -5,11 +5,9 @@ export
     LinearEquationOfState, RoquetIdealizedNonlinearEquationOfState,
     ∂x_b, ∂y_b, ∂z_b, buoyancy_perturbation, buoyancy_frequency_squared
 
-using 
-    Printf,
-    Oceananigans.Operators
-
-using Oceananigans: AbstractGrid
+using Printf
```
Comment (glwagner) [inline src/Buoyancy/Buoyancy.jl]: Is this preferred style? Why? I like fewer characters...

File: src/Grids/regular_cartesian_grid.jl [hunk 2]
```diff
@@ -41,7 +34,7 @@ struct RegularCartesianGrid{FT<:AbstractFloat, R<:AbstractRange} <: AbstractGrid
 end
 
 """
-    RegularCartesianGrid([FT=Float64]; size, length, x, y, z)
+    RegularCartesianGrid([FT=Float64]; size, length, topology, x, y, z)
```
Comment (glwagner) [inline src/Grids/regular_cartesian_grid.jl]: Hmm, the docstring does not contain the defaults. I think it should?

File: src/TimeSteppers/TimeSteppers.jl [hunk 3]
```diff
@@ -9,18 +9,15 @@ using GPUifyLoops: @launch, @loop, @unroll
...
-using Oceananigans.SurfaceWaves: x_curl_Uˢ_cross_U, y_curl_Uˢ_cross_U, z_curl_Uˢ_cross_U,
-                                 ∂t_uˢ, ∂t_vˢ, ∂t_wˢ
-
 using Oceananigans.Architectures: @hascuda
 @hascuda using CUDAnative, CuArrays
 
 using Oceananigans.Architectures
+using Oceananigans.Grids
 using Oceananigans.Operators
 using Oceananigans.Coriolis
```
Comment (glwagner) [inline src/TimeSteppers/TimeSteppers.jl]: Explicit imports might be nice here for all these physics terms... but we can save this for later.

**Pull request merged**